### PR TITLE
Fix multi-frame Zstandard response decoding

### LINKED ISSUE #3008: Partial responses when using chunked encoding + zstd compression
### Subject

I am testing using zstd compression with [bazel-remote](https://github.com/buchgr/bazel-remote/). As background, bazel-remote provides content-addressible storage under the /cas/ folder and storage of encoded build action data under the /ac/ folder. For the purposes of this bug report, I am only doing /cas/ downloads. So bazel-remote is used for basic PUT and GET of binary data.

Without providing the server headers to say that I support zstd compression, I don't receive chunked encoding and response.data contains the full binary data that I would expect. When I provide the server headers to say that I support zstd compression, the response from the server indicates chunked encoding and len(response.data) is always set to 1048576 bytes for files that are larger than that.

### Environment

OS Windows-10-10.0.22621-SP0
Python 3.10.11
urllib3 2.0.1

pip freeze results:

appdirs==1.4.4
certifi==2022.9.24
charset-normalizer==2.1.1
colorama==0.4.6
flake8==5.0.4
idna==3.4
mccabe==0.7.0
numpy==1.23.4
pycodestyle==2.9.1
pydiffx==1.1
pyflakes==2.5.0
python-gitlab==3.12.0
RBTools==4.0
requests==2.28.1
requests-toolbelt==0.10.1
six==1.16.0
texttable==1.6.7
tqdm==4.64.1
typing_extensions==4.4.0
urllib3==2.0.1
zstandard==0.21.0

### Steps to Reproduce

1. I deployed bazel-remote and did a PUT of a file larger than 1048576 bytes into /cas/<sha_256_hash_of_the_file_contents>
2. I ran the following code:
```
import urllib3

server_url = 'http://<server_name_redacted>:8000'
cas_url = '/cas/d1dcece4b328f7a12d86dd77cb0373496da2dd3971db2f3e3a29612f831a9738'
headers = urllib3.make_headers(accept_encoding=True)

response = urllib3.request('GET', server_url + cas_url, headers=headers)

print('Status: %d' % response.status)
print('Info: %s' % response.info())
print('Got %d back' % len(response.data))
```
### Expected Behavior

Script should print: Got 1308160 back.

### Actual Behavior

Script prints: Got 1048576 back

### Notes

If I set headers to None, the script outputs the following:
```
Status: 200
Info: HTTPHeaderDict({'Content-Length': '1308160', 'Content-Type': 'application/octet-stream', 'Date': 'Tue, 02 May 2023 12:47:40 GMT'})
Got 1308160 back
```
If I initialize headers to support compression, the script outputs the following:
```
Status: 200
Info: HTTPHeaderDict({'Content-Encoding': 'zstd', 'Content-Type': 'application/octet-stream', 'Date': 'Tue, 02 May 2023 12:48:24 GMT', 'Transfer-Encoding': 'chunked'})
Got 1048576 back
```

### PR DESCRIPTION
**Pull request by Rogdham**
Fixes #3008.

## Background on Zstandard

Data compressed with Zstandard is a sequence of frames. A frame can be of two kinds:
- Zstandard frame, containing compressed data
- Skippable frame, to be ignored by the decompressor

When decompressing some data, data from each frame should be concatenated.

[Link to the Zstandard specification](https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md)

## Multi-frame with python-zstandard

The [python-zstandard](https://python-zstandard.readthedocs.org) library used to decompress the data currently stops after the first frame. As a result, if several frames are chained in the response, the decompressed response will be partial.

This seems to be a temporary state of affairs as the author's library wants to introduce a `read_across_frames` parameter to their decompress APIs, which would default to `False` for some time before defaulting to `True`.

With the current version, I relied of the `unused_data` attribute of the `ZstdDecompressionObj` when `eof` is reached to continue with the next frame. This behaviour is specified [in the documentation](https://python-zstandard.readthedocs.io/en/latest/decompressor.html#zstddecompressionobj).

### CODE REVIEW

**Comment by grossag** [comment]
This solution is better than my https://github.com/urllib3/urllib3/pull/3021 . I like how you were able to create one `ZstdDecompressionObj` per frame. That's a clever way to work around the API's single frame limitation.

File: src/urllib3/response.py
Comment (grossag) [inline src/urllib3/response.py:173]: Is there any risk of an infinite loop if there is less data here than a frame's worth? Basically, is there a case where `decompress(unused_data)` results in more unused data because there is not enough data for zstd to do anything with?

File: src/urllib3/response.py
Comment (grossag) [inline src/urllib3/response.py:172]: I'm slightly concerned about decompress performance in this case. The current design of urllib3 seems to be to just push any data it has right to zstd and ignore `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE`, but won't that hurt performance when downloading larger files? It's nothing new to your change, but I am wondering if there is a solution that could respect `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` to not call `decompress()` so often.

**Review by grossag — COMMENTED** [review]
Clever solution! I like it, but have a few smaller questions.

File: src/urllib3/response.py
Comment (Rogdham) [reply src/urllib3/response.py:173]: My understanding is that in that case `eof` will be `False`

Comment (Rogdham) [reply src/urllib3/response.py:173]: This is confirmed with some testing:

```py
>>> compressed = zstd.compress(b'foo') + zstd.compress(b'bar')
>>> obj = zstd.ZstdDecompressor().decompressobj()

>>> obj.decompress(compressed[:10])
b'f'
>>> obj.eof
False
>>> obj.unused_data
b''

>>> obj.decompress(compressed[10:])
b'oo'
>>> obj.eof
True
>>> obj.unused_data
b'(\xb5/\xfd \x03\x19\x00\x00bar'
```

**Comment by Rogdham** [comment]
Hello @grossag thank you for your review!

I did not know you were working on it since you were not assigned to the ticket and your PR was not published when I started working on it. I'm glad you like my implementation though!

File: src/urllib3/response.py
Comment (Rogdham) [reply src/urllib3/response.py:172]: To test this I have created a 1GB file containing fully-random bytes (so that output is >1GB) and compressed it:
```
$ zstd -l data.zst
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
     1      0     954 MiB                       XXH64  data.zs
```

Then I run the following code with `pytest-benchmark`:

<details>
  <summary>Benchmark code</summary>
  
```python
import zstandard as zstd

with open("data.zst", "rb") as f:
    DATA = f.read()


def oneshot():
    obj = zstd.ZstdDecompressor().decompressobj()
    return len(obj.decompress(DATA))


def test_oneshot(benchmark):
    size = benchmark(oneshot)
    assert size == 1_000_000_000


def incremental():
    obj = zstd.ZstdDecompressor().decompressobj()
    size = 0
    for i in range(0, len(DATA), zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE):
        size += len(
            obj.decompress(DATA[i : i + zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE])
        )
    return size


def test_incremental(benchmark):
    size = benchmark(incremental)
    assert size == 1_000_000_000
```

</details>

The results show that decompressing one 1G string is ~2.5 time slower than splitting it and sending the data incrementally:

```
---------------------------------------- benchmark: 2 tests ----------------------------------------
Name (time in ms)          Min                 Max                Mean              Median          
----------------------------------------------------------------------------------------------------
test_incremental      317.5479 (1.0)      384.3817 (1.0)      358.2648 (1.0)      359.8864 (1.0)    
test_oneshot          921.7404 (2.90)     968.6641 (2.52)     949.1417 (2.65)     964.4232 (2.68)   
----------------------------------------------------------------------------------------------------
```

---

Now I leave urllib3's maintainer decide if they want me to make a quick change in the code to do so or if that's ok as it is.

One could also argue that python-zstandard should do it automatically.

Comment (Rogdham) [reply src/urllib3/response.py:172]: Also @grossag I did not find much documentation about `DECOMPRESSION_RECOMMENDED_INPUT_SIZE` (nor  `ZSTD_DStreamInSize()` for that matter), feel free to share if you have some poniters!

**Comment by sethmlarson** [comment]
Thank you @grossag and @Rogdham  for all your work on this issue, I'd like to invite you both to submit an Open Collective invoice of $300 upon completion if you're able to! 🙏 

I'm taking a look at this PR approach tonight.

File: src/urllib3/response.py
Comment (grossag) [reply src/urllib3/response.py:172]: `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` defaults to the result of a call into zstd's function `ZSTD_DStreamInSize()`. After reading through some zstd source and also https://github.com/facebook/zstd/issues/340 , I think this value may represent the chunk size that zstd performs best with.

That said, optimizing for network download + decompress is much different than disk read + decompress. Also, the variance will be much higher in terms of network download speed. So it's hard to know how much to tune this.

The downside with delaying decompression until you have a buffer of size `DECOMPRESSION_RECOMMENDED_INPUT_SIZE` is that it defers more work into `flush()`, which as shown in https://github.com/urllib3/urllib3/pull/3021 can break tests. If we delay decompression, the normal case for smaller chunked transfers is that `flush()` will actually do the decompress operation.

So I am fine either way and will defer to @sethmlarson on whether more work is needed to delay decompression. Either way, I still think we should go with your approach. Mine has the benefit of using the only API that is known to be able to handle multiple frames now, which could slightly improve performance in that case. But I think the performance gain is lost by my approach having to use an intermediate `io.BytesIO()` which would result in an additional copy. I like how your approach doesn't need extra copies.

File: src/urllib3/response.py
Comment (grossag) [reply src/urllib3/response.py:173]: Actually, sorry, one more question here. What if there is truncated data? Is there any case where there would repeatedly be unused data if the network connection was closed early without receiving everything? Or would it throw an exception in that case?

Comment (Rogdham) [reply src/urllib3/response.py:173]: If the `eof` attribute is True and there is some `unused_data`, then it will go into the next loop. At that point a new object is instantiated, and the `unused_data` from the previous one is fed for decompression. And so on and so forth.

If you are concerned about possible infinite loop, then it will not happen because the `unused_data` part between each iteration of the loop is guarantied to strictly decrease in size with each iteration.

Now, if the data is truncated, then nothing special will happen. The part of the received data that can be decompressed will be returned, and that will be it. No exception is raised in that case, at least due to the code about Zstandard decompression; I would guess that the code from HTTP handling will detect that packets are lost or something like this and react accordingly.

Not sure if that answers your question, please feel free to clarify otherwise!

File: src/urllib3/response.py
Comment (Rogdham) [reply src/urllib3/response.py:172]: Thank you for the pointers.

I am confused though. I though you asked about what happens when we feed **more** data than `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` to the decompressor, but then in your last message you speak about what happens when we feed **less** data. So here are my thoughts:

- Feeding more data: we see in the benchmark that there seems to be a penalty cost to do so (600ms per GB). I'm open to changing the code to consider that but would argue that the python-zstandard's lib should do it itself.
- Feeding less data: this case allows for the consumer of urllib3 to get the data as soon as it is ready, which would be helpful when streaming HTTP responses. I'm not sure how often this scenario would happen in real life though.

Also, you spoke about calling `flush()`, but in the case of this PR `ZstdDecompressionObj.flush` does nothing at all. I put a comment in the code for clarity, but did not remove the code calling it in case python-zstandard's API changes in the future.

Comment (grossag) [reply src/urllib3/response.py:172]: Sorry! I am wondering specifically about feeding less data than `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE`. But as discussed in Discord, zstd frames can be any size so my concern isn't important.

My biggest concern was: what happens when we receive half of a frame's worth of data? python-zstandard's underlying call to zstd's `ZSTD_decompressStream` will return 3, meaning that there is more data to decode. But there isn't enough data to decode so we basically need to wait for more data.

I wrote some Python code where I read the full contents of a .zstd file and sliced it in half at different points. I then ran your code over it. It worked perfectly for various ways that I sliced the buffer, so I think your approach is great.

The part I didn't realize is that zstd internally remembers where it is at during a decompress operation. If there is not enough data, it will do as much as it can and return. Then your code is doing the right thing in reusing `self._obj` if it doesn't have enough data and it's mid-frame (`self._obj.eof` being False if we are mid-frame).

File: src/urllib3/response.py
Comment (grossag) [reply src/urllib3/response.py:173]: Yes, that answers my question, thanks!

File: src/urllib3/response.py
Comment (Rogdham) [reply src/urllib3/response.py:172]: Great! Thank you for taking the time to confirming that everything works as expected! :+1:

**Comment by grossag** [comment]
This change looks good to me. I tested thousands of zstd-compressed GETs with your topic branch and confirmed that they all worked correctly.

I am not a maintainer of urllib3 so I am not sure if I should hit the Approve button. I'll defer to the maintainers on approving it. But I think it's a great improvement that will help a lot, thanks for your work on this!

**Review by pquentin — APPROVED** [review]
This looks great, thanks! Did you check that the test was failing without your change?

I left two small questions.

File: test/test_response.py
Comment (Rogdham) [reply test/test_response.py:341]: The magic number is in little-endian, and the last half-byte (noted `?` in the spec) can be anything, I chose `0` here but it is arbitrary. So 0x184D2A50, which encoded little-endian is `502A4D18`.

Then followed by the frame size, I chose 16, so `10000000`.

Then the user data part, I used 16 arbitrary bytes.

Maybe it would be clearer if I put some spaces in there, like one of the following:
```python
bytes.fromhex("502A4D18 10000000 726f676468616d207761732068657265")
bytes.fromhex("50 2A 4D 18    10 00 00 00    72 6f 67 64 68 61 6d 20 77 61 73 20 68 65 72 65")
```

Tell me if that makes more sense or if you want me to change it to something else.

**Comment by Rogdham** [comment]
> Did you check that the test was failing without your change?

The failing test is a timeout in one of the environments only because it reached 30min; I don't think it's related to my change.

**Comment by grossag** [comment]
> > Did you check that the test was failing without your change?
> 
> The failing test is a timeout in one of the environments only because it reached 30min; I don't think it's related to my change.

I think the question was whether your new zstd test fails without your current change. And I did confirm previously when reviewing your change that the new zstd test fails when run on current main top of tree. So you should be all set for that.

**Comment by Rogdham** [comment]
> I think the question was whether your new zstd test fails without your current change.

Oh my bad I completely misread the question. Thank you for clarifying (and having tested it!), I have also confirmed it on my side.

File: src/urllib3/response.py
Comment (pquentin) [reply src/urllib3/response.py:180]: OK, the first reason makes sense, thanks.

File: test/test_response.py
Comment (pquentin) [reply test/test_response.py:341]: I like option 3! Can we mention additionally that magic number and frame size are little-endian? And maybe use only zeroes for user data? Otherwise it sounds like the randomness is significant.

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks, I should have been clearer regarding the failing test. The only thing remaining is commenting the skippable frame per https://github.com/urllib3/urllib3/pull/3022#discussion_r1191039333

File: test/test_response.py
Comment (sethmlarson) [reply test/test_response.py:341]: I agree, let's add comments to the parts of the frame and use zeroes to denote insignificant data.

**Review by sethmlarson — CHANGES_REQUESTED** [review]
This looks great, thanks for working on this! I had two comments:

File: test/test_response.py
Comment (Rogdham) [reply test/test_response.py]: So to test I wrote a quick TCP server to send the same exact data as request body.

<details><summary>HTTP server code</summary>
<p>

```python
import socketserver
import zstandard as zstd


data = (
    # Zstandard frame
    zstd.compress(b"foo")
    # skippable frame (must be ignored)
    + bytes.fromhex(
        "50 2A 4D 18"  # Magic_Number (little-endian)
        "07 00 00 00"  # Frame_Size (little-endian)
        "00 00 00 00 00 00 00 "  # User_Data
    )
    # Zstandard frame
    + zstd.compress(b"bar")
)


class MyTCPHandler(socketserver.BaseRequestHandler):
    def handle(self):
        f = self.request.makefile("rb")
        while f.readline().strip():
            pass  # consume request
        self.request.sendall(
            (
                b"HTTP/1.1 200 OK\r\n"
                b"Content-Encoding: zstd\r\n"
                b"Content-Type: text/plain\r\n"
                b"Content-Length: %d\r\n"
                b"\r\n"
                b"%s"
            )
            % (len(data), data)
        )


if __name__ == "__main__":
    HOST, PORT = "localhost", 9999
    with socketserver.TCPServer((HOST, PORT), MyTCPHandler) as server:
        server.serve_forever()
```

</p>
</details> 

And then queried it with curl:
```sh
$ curl --compressed http://localhost:9999/
foobar
```

However both Firefox & Chromium directly download the compressed file. It's not clear to me but it seems that [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1301878) and [Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=1246971) don't support zstd compression yet.

And confirmed the fix of this PR:
- with `urllib3==2.0.2`:
```python
>>> urllib3.request('GET', 'http://localhost:9999').data
b'foo'
```
- with the PR:
```python
>>> urllib3.request('GET', 'http://localhost:9999').data
b'foobar'
```

Comment (Rogdham) [reply test/test_response.py]: Done! I have changed the frame size to 7 so that it is clearer that it does not need to be a round number.

Comment (sethmlarson) [reply test/test_response.py]: curl doing the right thing is enough for me, thank you!

**Review by sethmlarson — DISMISSED** [review]
LGTM, thanks @Rogdham and @grossag!

File: test/test_response.py
Comment (pquentin) [inline test/test_response.py:344]: nit: Can you please remove the trailing space after the last 00?

**Review by pquentin — DISMISSED** [review]
LGTM! Thanks for iterating. And thanks everyone for the careful reviews.

File: test/test_response.py
Comment (Rogdham) [reply test/test_response.py]: Sorry for that, fixed!

**Review by pquentin — APPROVED** [review]

**Review by sethmlarson — APPROVED** [review]

**Comment by sethmlarson** [comment]
Thanks @Rogdham and @grossag! You are both invited to apply for the $300 OpenCollective bounty. Please submit an expense to [our OpenCollective](https://opencollective.com/urllib3) if you're eligible to do so and let us know if there's anything we can do to help with that process :)

**Comment by grossag** [comment]
> Thanks @Rogdham and @grossag! You are both invited to apply for the $300 OpenCollective bounty. Please submit an expense to [our OpenCollective](https://opencollective.com/urllib3) if you're eligible to do so and let us know if there's anything we can do to help with that process :)

Thanks @sethmlarson ! I'm not eligible but appreciate the offer. The fix is going to be very useful for my work anyway. Thanks @Rogdham for making it.

**Comment by Rogdham** [comment]
Thank you for merging the PR, and for the bounty offer! I have sent [an expense](https://opencollective.com/urllib3/expenses/138985) on OpenCollective. Feel free to contact me if you have any questions with this piece of code in the future. Many thanks!

**Comment by pquentin** [comment]
Thanks! Approved. ✅

**Pull request merged**
